python/pyarrow/tests/test_parquet.py

```diff
@@ -2028,13 +2041,39 @@ def test_write_error_deletes_incomplete_file(tempdir):
 
     filename = tempdir / 'tmp_file'
     try:
-        _write_table(pdf, filename)
+        _write_table(pdf, filename, coerce_timestamps='ms')
     except pa.ArrowException:
         pass
 
     assert not filename.exists()
 
 
+@pytest.mark.pandas
+def test_noncoerced_nanoseconds_written_without_exception(tempdir):
+    # ARROW-1957
+    n = 9
+    df = pd.DataFrame({'x': range(n)},
+                      index=pd.DatetimeIndex(start='2017-01-01',
+                      freq='1n',
+                      periods=n))
+    tb = pa.Table.from_pandas(df)
+
+    filename = tempdir / 'written.parquet'
+    try:
+        pq.write_table(tb, filename)
+    except Exception:
+        pass
+    assert filename.exists()
+
+    recovered_table = _simple_table_roundtrip(tb)
+    assert tb.equals(recovered_table)
+
+    # Loss of data thru coercion (without explicit override) still an error
+    filename = tempdir / 'not_written.parquet'
+    with pytest.raises(ValueError):
+        pq.write_table(tb, filename, coerce_timestamps='ms')
+
+
```

Comment:
So this is a nuisance, but old Parquet readers (i.e. most deployed Big Data applications) aren't going to be able to read the new nanosecond metadata -- for example, most deployed versions of Spark! For version `1.0` files, I think the appropriate default behavior is to write in compatibility mode in millis or micros with the old ConvertedType set, while for `version=2.0` I think writing the new Nanosecond annotation is acceptable.

cc @xhochy for any comments

Reply:
OK, so I've reintroduced the old behavior for timestamps if you've got a v1.0 writer; mostly this means that (1) Arrow nanos timestamps are converted to micros by default, and (2) you can't coerce any timestamp to nanoseconds (unless you use int96's).  Conformed this and other tests and added some new tests to demonstrate and check these as well.

Can I ask for clarification on something: recall that (after PARQUET-1411) Thrift serialized Parquet schemas always contain *both* ConvertedType and LogicalType (aka annotation) information (modulo types that don't mutually convert).  Are you saying here that for version 1.0 writers (in addition to the Arrow specific change I described above) that the Parquet library proper should not even serialize the LogicalType/annotation?

Reply:
One further note: the pre-existing design has us asking and answering the same non-trivial questions about timestamp typing twice: once when the schema is built (GetTimestampMetadata()) and again when the data arrays are built (ArrowColumnWriter::WriteTimestamps()).  I retained this design, but it's straining under this patch.  I believe that after a schema is built, we have all the information needed to provide direct instructions to the column/array writer, but the module doesn't provide any channel for that (at least not an obvious one -- I haven't spent too much time analyzing it).  I think a refactor for this would be desirable, but is probably beyond the scope of what I can do right now (given the desire to get 0.14 out in a few weeks, and the "big rename" for ConvertedType still pending and blocked).  Just wanted to flag this for you.

Reply:
Agreed that the current structure isn't ideal, and a refactor is a good idea